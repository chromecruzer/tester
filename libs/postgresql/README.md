# @trac-postgresql

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test @trac-postgresql` to execute the unit tests via [Jest](https://jestjs.io).
