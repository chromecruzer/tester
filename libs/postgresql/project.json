{
  "name": "postgresql",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/postgresql/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/postgresql/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/libs/postgresql"],
      "options": {
        "jestConfig": "libs/postgresql/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
