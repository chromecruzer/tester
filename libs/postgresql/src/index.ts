export * from './postgresql';
import PostgresConfigFactory, {PostgresConfig} from "./PostgresConfigFactory";
import BulkSqlWithCreate from "./BulkSqlWithCreate";
export *  from './ErrorHandler';
import ExtractIolData from "./ExtractIolData";
import PostgresClient from "./PostgresClient";
import LoadXlsxData from './LoadXlsxData';
import NotificationsServer from "./NotificationsServer";
import SearchTable from "./SearchTable";
import ImportIntoConsignments from "./ImportIntoConsignments";
import ImportIntoJumpTable from "./ImportIntoJumpTable";
import AuditStore from "./AuditStore";
import UserStore from './UserStore';

export {
  AuditStore, PostgresConfig, PostgresConfigFactory, BulkSqlWithCreate, ExtractIolData,
  PostgresClient, LoadXlsxData, NotificationsServer, SearchTable, ImportIntoConsignments, ImportIntoJumpTable,
  UserStore
};
