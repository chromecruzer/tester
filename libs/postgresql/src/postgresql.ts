import * as _ from "lodash";
import {DateTime} from "luxon";
import {PoolClient} from "pg";
import util from "util";
// eslint-disable-next-line @nrwl/nx/enforce-module-boundaries
import {
  customerContactDataMapping,
  customerDataMapping,
  dataTableNames, InputDataType,
  InputDataTypeLabel,
  InputToSqlMapping,
  iolDataMapping,
  matchCustomerContactFields,
  matchCustomerFields,
  matchIolFields,
  matchProductFields,
  matchSalesMappingFields,
  matchSalesRepFields,
  NullableString,
  productDataMapping,
  employeeRecordDm,
  salesMappingDataMapping,
  salesRepDataMapping,
  tracDateFormat, 
  UploadRecord,
  matchEmployeeFields,
  userRecordDm,
  matchUser1Fields
} from "../../datatypes/src";
import path from "path";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

export const auditTableNames = {
  notes: 'auditnotes',
  locations: 'auditlocation',
  items: 'audititem',
}



export const searchTableNamesFn = () => [
  dataTableNames.consignments,
  dataTableNames.customers,
  dataTableNames.customerContacts,
  dataTableNames.products,
  dataTableNames.salesReps];

export const sqlDate = (datum: (Date | string)) => {
  // console.log(`datum ${dump(datum)}
  // from JSDate ${dump(DateTime.fromJSDate(datum))}
  // to format ${dump(DateTime.fromJSDate(datum).toFormat(tracDateFormat))}`)
  const date = _.isDate(datum) ? datum : new Date(datum);
  return DateTime.fromJSDate(date).toFormat(tracDateFormat)
}

export function datumToSql(datum: (string | number | Date | boolean), datatype: InputDataTypeLabel) {
  if (_.isNull(datum) || _.isUndefined(datum) || datum === '') {
    return 'NULL';
  }
  let str;
  switch (datatype) {
    case 'string':
      str = String(datum);
      if (str.length === 0) {
        return null;
      }
      str = str.replace(/'/g, "''");
      str = str.replace(/\n/g, "\\n");
      return `'${str}'`;
    case 'date':
      // console.log(`date ${datum} converted ${sqlDate(datum as unknown as Date)}`)
      return `'${sqlDate(datum as Date)}'`
    case 'boolean':
      // console.log(`boolean ${datum} as ${Boolean(datum)}`)
      return Boolean(datum);
    case 'number':
      return datum;
  }
}

export function appendUuids(sql: string, uuids: string[]) {
  let skipped1st = false;
  let result = sql;
  uuids.forEach(id => {
    result += `${skipped1st ? ',' : ''}    \n'${id}'`;
    skipped1st = true;
  });
  result += ');';
  return result;
}

export const checkModified = modified => {
  // console.log(`Checking ${dump(modified)}`);
  if (modified.originalUuid !== modified.new.uuid) {
    console.error(`UUID mismatch found ${dump(modified)}`);
  }
}

export function appendBulkData(sql, bulkData: UploadRecord[],
                               columnMapping: InputToSqlMapping[],
                               createUuid = true,
                               hasUuid = false) {
  let skipped1stRow = false;
  let result = `${sql}\n`;
  bulkData.forEach((record) => {
    // console.log(`append bulk data ${dump(record)}`)
    const dataRow = [];
    if (createUuid) {
      dataRow.push('gen_random_uuid()')
    }
    if(hasUuid) {
      dataRow.push(datumToSql(record['uuid'], 'string'));
    }
    columnMapping.forEach((colMap) => {
      let datum = datumToSql(record[colMap.sqlLabel], colMap.dataType);
      if(!_.isUndefined(colMap.default) && datum === null) {
        datum = colMap.default;
      }
      dataRow.push(datum)
    });
    result += `${skipped1stRow ? ',': ''}\n(${_.join(dataRow, ', ')})`;
    skipped1stRow = true;
  });
  return result;
}

export const queryRecords = async (client: PoolClient, recordSql: string, idField: string,
                                   {
                                     uuid = null as NullableString,
                                     id = null as NullableString,
                                     ids = [] as string[],
                                   }) => {
  let sql = recordSql;
  const check = (_.isString(uuid) ? 1 : 0) + (_.isString(id) ? 1 : 0) + (_.isEmpty(ids) ? 0 : 1);
  if (check > 1) {
    throw new Error(`cannot set multiple parameters ${uuid}, ${id}, ${ids.length}`);
  }
  if (uuid) {
    sql += ` WHERE uuid='${uuid}'`;
  }
  if (id) {
    sql += ` WHERE ${idField}='${id}'`;
  }
  if (!_.isEmpty(ids)) {
    sql += ` WHERE ${idField} IN (${_.join(_.map(ids, i => `'${i}'`), ', ')})`;
  }
  sql += ';';
  // console.log(`query records sql '${sql}'`);
  return client.query(sql)
    .then(data => data.rows)
    .catch(err => {
    throw new Error(`Mistake in '${sql}' caused ${err.message}`);
  });
}

interface TransformerKeys {
  matchFields: (string | string[]);
  matchFn: (a: any, b: any) => boolean,
  mappings: InputToSqlMapping[];
  compareFieldList?: string[],
}

export const recordFromArray = (row: InputDataType[], mapping: InputToSqlMapping[]) => {
  return _.reduce(mapping, (accum, m,i) => {
    const r = row[i] || null;
    let value = _.isFunction(m.nullify) ? m.nullify(r) : r;
    if(_.has(m, 'default')) { //need to us _.has here because the default may be a boolean false
      // console.log(`checking default for ${dump(m)} with value ${value}`)
      value = value || m.default;
    }
    accum[m.sqlLabel] = value;
    return accum;
  }, {} as UploadRecord);
}
export const transformHelpers: { [id: string]: TransformerKeys } = {
  iol: {
    matchFields: 'lot',
    matchFn: matchIolFields,
    mappings: iolDataMapping,
  },
  customers: {
    matchFields: 'customer_code',
    matchFn: matchCustomerFields,
    mappings: customerDataMapping,
  },
  customer_contacts: {
    matchFields: ['full_name', 'ship_to_id'],
    matchFn: matchCustomerContactFields,
    mappings: customerContactDataMapping,
  },
  products: {
    matchFields: 'product_id',
    matchFn: matchProductFields,
    mappings: productDataMapping,
  },
  employee: {
     matchFields: 'emp_id',
     matchFn: matchEmployeeFields,          // added by shyam for testing
     mappings: employeeRecordDm,
  },

  users: {
    matchFields: 'username',
    matchFn:  matchUser1Fields,              // added by shyam for testing
    mappings: userRecordDm
  },
  salesmappings: {
    matchFields: 'customer_ship_to',
    matchFn: matchSalesMappingFields,
    mappings: salesMappingDataMapping,
    compareFieldList: ['customer_ship_to', 'postal', 'zipcode', 'primary_territory', 'primary_name', 'cos_territory',
      'cos_name', 'kae_territory', 'kae_name'],
  },
  salesreps: {
    matchFields: 'salesrep_id',
    matchFn: matchSalesRepFields,
    mappings: salesRepDataMapping,
  },
}
export const matchFieldFn = (fields, record) => {
  return _.isArray(fields) ? _.join(_.map(fields, f => record[f]), '|') : record[fields]
}
export const deepEqualFn = (c, m, compareFieldList = null) => {
  if (_.isArray(compareFieldList)) {
    return _.reduce(compareFieldList,
      (boolAccum, cf) => {
        if (!boolAccum) {
          return boolAccum;
        }
        // console.log(`comparing isequal field ${cf}, ${candidate[cf]} to ${m[cf]}`);
        return _.isEqual(c[cf], m[cf]);
      }, true);
  }
  return _.isEqual(m, c);
}
export const docxTemplateFn = (apiConfig, templateType) => path.join(apiConfig.templatesPath,
  (templateType === 'Missing' ? 'missingTemplate.docx' : 'expiredTemplate.docx'))
export const htmlTemplateFn = (apiConfig, templateType) => path.join(apiConfig.templatesPath,
  (templateType === 'Missing' ? 'missingTemplate.html' : 'expiredTemplate.html'))

