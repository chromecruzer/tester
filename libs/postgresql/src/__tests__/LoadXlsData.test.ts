import {LoadXlsxData} from "@trac/postgresql";
import {Settings} from "luxon";
import _ from "lodash";

describe('LoadXlsData', () => {
  let fs, xlsx, loader;
  beforeEach(() => {
    fs = {
      readFile: jest.fn(),
    };
    xlsx = {
      read: jest.fn(),
    }
    loader = new LoadXlsxData();
    loader.fs = fs;
    loader.xlsx = xlsx;
  });
  describe('readWithMap', () => {
    let sheets, mapping;
    beforeEach(() => {
      Settings.defaultZone = 'Central Time'
      mapping = [
        {
          xlsxAddress: 'A',
          dataType: 'string',
          sqlLabel: 'sqlLabelString',
          sqlType: 'char(5)'
        },
        {
          xlsxAddress: 'B',
          dataType: 'number',
          sqlLabel: 'sqlLabelNumber',
          sqlType: 'integer'
        },
        {
          xlsxAddress: 'C',
          dataType: 'date',
          sqlLabel: 'sqlLabelDate',
          sqlType: 'date'
        }
      ];
      sheets = {
        SheetNames: ['IOL'],
        Sheets: {
          'IOL': {
            'A1': {
              w: 'Should not be in snapshot',
            },
            'B1': {
              w: 'Should not be in snapshot',
            },
            'C1': {
              w: 'Should not be in snapshot',
            },
            'A2': {
              w: 'I am a string',
            },
            'B2': {
              w: 57,
            },
            'C2': {
              w: '10/10/2021',
            },
            'A3': {
              w: 'I am also a string',
            },
            'B3': {
              w: 123,
            },
            'C3': {
              w: '10/11/2021',
            },
            'A4': {
              w: 'I am still a string',
            },
            'B4': {
              w: 2001,
            },
            'C4': {
              w: '12/10/2021',
            },
          }
        }
      };
    });
    afterEach(() => {
      Settings.defaultZone = 'system'
    });
    it('should return an error if the XLSX file does not exist', () => {
      const expected = new Error('file does not exist');
      fs.readFile.mockImplementation(() => Promise.reject(expected));
      xlsx.read.mockReturnValue(sheets);
      return expect(loader.readWithMap('I am not a path', mapping)).rejects.toMatchSnapshot();
    });
    it('should return an error if the XLSX file does not parse', () => {
      const expected = new Error('XLSX is giving up');
      fs.readFile.mockImplementation(() => Promise.resolve('I am a buffer'));
      xlsx.read.mockImplementation(() => {
        throw expected;
      });
      return expect(loader.readWithMap('I am a bad file', mapping)).rejects.toEqual(expected);
    });
    it('should handle parsing of strings, dates, and numbers', () => {
      fs.readFile.mockImplementation(() => Promise.resolve('I am a buffer'));
      xlsx.read.mockReturnValue(sheets);
      return loader.readWithMap('I am a real xlsx file', mapping).then(() => {
        // console.log('checking bulk data');
        expect(loader.bulkData).toMatchSnapshot();
      })
    });
  });
  describe('scanForLotNumbers', () => {
    const createAuditScan = (includeHeader) => {
      const sheets = {
        SheetNames: ['Sheet1'],
        Sheets: {
          'Sheet1': null,
        }
      };
      const dataCells = {
        'A1': {
          w: '01234567',
          v: 1234567,
          t: 'n'
        },
        'B1': {
          w: 'Text',
          v: 'Text',
          t: 's'
        },
        'C1': {
          w: 'true',
          v: true,
          t: 'b'
        },
        'D1': {
          w: '10/10/2021',
          t: 'd'
        },
        'A2': {
          w: '08901234',
          v: 8901234,
          t: 'n'
        },
        'A3': {
          w: '08901235',
          v: '08901235',
          t: 's'
        },
        'B3': {
          w: 'Number as text',
          v: 'Number as text',
          t: 's'
        },
      };
      if (includeHeader) {
        const headerCells = {
          'A1': {
            w: 'S/N',
            v: 'S/N',
            t: 's'
          },
          'B1': {
            w: 'Text',
            v: 'Text',
            t: 's'
          },
          'C1': {
            w: 'Boolean',
            v: 'Boolean',
            t: 's'
          },
          'D1': {
            w: 'Date',
            v: 'Date',
            t: 's'
          },
        };
        const cellRegex = /([A-Z])(\d+)/;
        sheets.Sheets['Sheet1'] = _.reduce(dataCells, (accum, val, cellA) => {
          const [cell, col, row] = cellRegex.exec(cellA)
          // console.log(`cell ${cell} row ${row} col ${col}`);
          const fixed = Number(col) + 1;
          accum[`${col}${row}`] = val;
          return accum
        }, headerCells);
        return sheets;
      }
      sheets.Sheets['Sheet1'] = dataCells;
      return sheets;
    }
    it('should scan one column no headers', () => {
      const sheets = createAuditScan(false);
      fs.readFile.mockImplementation(() => Promise.resolve('I am a buffer'));
      xlsx.read.mockReturnValue(sheets);
      return loader.scanForLotNumbers('path/to/xlsx/file.xlsx')
        .then(() => {
          expect(loader.lotNumbers).toMatchSnapshot();
        });
    });
    it('should ignore text rows', () => {
      const sheets = createAuditScan(true);
      fs.readFile.mockImplementation(() => Promise.resolve('I am a buffer'));
      xlsx.read.mockReturnValue(sheets);
      return loader.scanForLotNumbers('path/to/xlsx/file.xlsx')
        .then(() => {
          expect(loader.lotNumbers).toMatchSnapshot();
        });
    });
  });
});
