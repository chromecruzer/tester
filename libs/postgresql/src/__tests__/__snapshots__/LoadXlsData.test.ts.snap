// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadXlsData readWithMap should handle parsing of strings, dates, and numbers 1`] = `
Array [
  Array [
    "I am a string",
    57,
    2021-10-10T05:00:00.000Z,
  ],
  Array [
    "I am also a string",
    123,
    2021-10-11T05:00:00.000Z,
  ],
  Array [
    "I am still a string",
    2001,
    2021-12-10T06:00:00.000Z,
  ],
]
`;

exports[`LoadXlsData readWithMap should return an error if the XLSX file does not exist 1`] = `[Error: "I am not a path" could not be read because Error: file does not exist]`;

exports[`LoadXlsData scanForLotNumbers should ignore text rows 1`] = `
Array [
  "01234567",
  "08901234",
  "08901235",
]
`;

exports[`LoadXlsData scanForLotNumbers should scan one column no headers 1`] = `
Array [
  "01234567",
  "08901234",
  "08901235",
]
`;
